### ISSUE TITLE
Intellij 2024.2: Investigating unit test failure for 2024.2 intellij version

### ISSUE THREAD

**Issue by staicy123** [OpenLiberty/liberty-tools-intellij#1048]
After making changes in build.gradle file for supporting intellij 2024.2. When running the unit test for intellij 2024.2. The build is neither getting failed nor terminated for a long time. The termination takes more than 10 minutes. The same tests works fine for intellij 2024.1 version.
Logs: 
[log.txt](https://github.com/user-attachments/files/17486001/log.txt)

Unit tests failures are listed below after fix `importProjectsWithErrors(pomFiles.toArray(VirtualFile[]::new));`

- [x] io.openliberty.tools.intellij.lsp4jakarta.it.beanvalidation.BeanValidationTest#validFieldConstraints
- [x] io.openliberty.tools.intellij.lsp4jakarta.it.di.DependencyInjectionTest#DependencyInjectionDiagnostics
- [x] io.openliberty.tools.intellij.lsp4jakarta.it.persistence.JakartaPersistenceTest#removeFinalModifiers
- [x] io.openliberty.tools.intellij.lsp4jakarta.it.websocket.JakartaWebSocketTest#testDuplicateOnMessage
- [x] io.openliberty.tools.intellij.lsp4jakarta.it.websocket.JakartaWebSocketTest#addPathParamsAnnotation

**Comment by vaisakhkannan** [comment]
Link to the latest Jetbrains Slack discussion : https://jetbrains-platform.slack.com/archives/C5U8BM1MK/p1730119000848479?thread_ts=1730105939.390899&cid=C5U8BM1MK

**Comment by staicy123** [comment]
Hi @TrevCraw,

I would like to provide a detailed explanation of this issue. Based on the analysis, I found that the test fails when it reaches this method. Please find the relevant code and logs attached below.
<img width="869" alt="Screenshot 2024-10-30 at 12 33 18 PM" src="https://github.com/user-attachments/assets/c10e11ed-cbfc-40d6-9563-cfe21ffa60ef">
<img width="988" alt="Screenshot 2024-10-30 at 12 34 39 PM" src="https://github.com/user-attachments/assets/2f15b219-d20f-4e3c-8e86-bb3839db5922">

Through discussions in the JetBrains Slack channel, it became clear that the issue is caused by a transitive dependency on kotlinx-coroutines. JetBrains provides a testFramework() dependency helper that might resolve this issue.

The testFramework(TestFrameworkType.Platform), testFramework(TestFrameworkType.Plugin.Maven), and similar configurations – when applied – add dependencies on the requested library version closest to your IntelliJ Platform version. They also exclude transitive dependencies already present in the IntelliJ Platform archive, including coroutines.

If you add a dependency on testFramework directly using testImplementation, it will include all transitive dependencies. Disabling transitive dependencies manually (using transitive = false) won’t work, as some dependencies are actually required. That’s why the testFramework helper exists.

I removed the manual Maven dependency for testFramework and implemented the suggested helper framework for Maven. However, kotlinx-coroutines is still appearing as a dependency in our project. When I reported this to the JetBrains team, they confirmed it’s a bug on their end and plan to address it in the next release. I’ve created an issue to track this.
https://youtrack.jetbrains.com/issue/IJSDK-2276/Even-after-adding-testFrameworktype-helper-could-still-see-coroutines-in-transitive-dependencies-hence-unit-test-failing
https://github.com/JetBrains/intellij-platform-gradle-plugin/issues/1817

Please find the Slack conversation attached here: https://jetbrains-platform.slack.com/archives/C5U8BM1MK/p1730105939390899

They also confirmed that they will try to fix at the earliest 

Thanks

**Comment by staicy123** [comment]
Hi @TrevCraw, 

**Update on the Issue:**

As of today, we tested the snapshot version provided by the team on Friday evening. They attempted to fix the issue, but we confirmed that the problem still persists, and we are unable to run the unit tests using version 2024.2.

On their request, I shared the GitHub link where we are making changes to support version 2024.2 and am currently waiting for their reply.

Analysis:

From my findings, it appears that the kotlinx dependency has been removed in the new version. Our tests are still expecting this dependency to run successfully. I'm not entirely sure about this observation, but I would appreciate any contributions or insights from your end if you have any information on this matter.

Thanks

**Comment by TrevCraw** [comment]
Hi @staicy123 , I don't have any more insight into this. I think we need to continue to communicate with the JetBrains team and see what new information we get from them.

**Comment by vaisakhkannan** [comment]
Based on Jetbrains Replay


_I see you have_

`platformBundledPlugins=com.intellij.java, org.jetbrains.idea.maven, com.intellij.gradle, org.jetbrains.plugins.terminal, com.intellij.properties`

_Currently, Intellij Gradle plugin 2.1.0 has an issue with loading all required bundled modules when targeting 2024.2_


_The final 2.2.0 will handle that properly (I’m still waiting for Plugin Verifier libraries to be fixed)_



We need to wait for the version 2.2.0 and then need to test this unit test failure issue

**Comment by vaisakhkannan** [comment]
Since we are using IntelliJ version `2024.1.7` for the production build, we will ensure the unit tests are running. Therefore, I don’t think we need to wait for the JetBrains team to release a new Gradle plugin version, as they haven’t done so yet. Consequently, we cannot test this on IntelliJ 2024.2 with the latest Gradle plugin. Since we are targeting IntelliJ `2024.1.7` as the `production build` version for this release, we can remove this issue from our current milestone.

**Comment by vaisakhkannan** [comment]
The JetBrains team has released a new Gradle IntelliJ plugin, version 2.2.0 - https://github.com/JetBrains/intellij-platform-gradle-plugin/releases/tag/2.2.0. We should test this issue using the latest version and verify whether the unit tests fail in IntelliJ versions 2024.2 and 2024.3. Based on our previous discussion with them, they mentioned that this new Gradle plugin version might resolve our unit test failures.

**Comment by vaisakhkannan** [comment]
I tested with the new Gradle IntelliJ plugin, version 2.2.0, but I am still getting the unit test failure mentioned in this issue.

**Comment by vaisakhkannan** [comment]
I tried updating the version from `2024.1.7` to `2024.2` and built the project. However, I encountered same unit test failure as mentioned in the issue #1048. The failure pointed to a specific [line](https://github.com/OpenLiberty/liberty-tools-intellij/blob/672fbcb09d6c185f9bf4a38df15f2ffc2b9f190d/src/test/java/io/openliberty/tools/intellij/lsp4jakarta/it/core/BaseJakartaTest.java#L80C9-L83C11) of code located in the `BaseJakartaTest` class, which extends the MavenImportingTestCase class. The line invokes the method `importProjectsAsync(pomFiles.toArray(VirtualFile[]::new), continuation)`, which is defined in the `MavenImportingTestCase` class.


```
protected suspend fun importProjectsAsync(vararg files: VirtualFile) {
    importProjectsAsync(listOf(*files))
  }

  protected open suspend fun importProjectsAsync(files: List<VirtualFile>) {
    if (preimportTestMode) {
      val activity = ProjectImportCollector.IMPORT_ACTIVITY.started(project)
      try {
        MavenProjectStaticImporter.getInstance(project)
          .syncStatic(files, null, mavenImporterSettings, mavenGeneralSettings, true, SimpleStructureProjectVisitor(), activity, true)
      }
      finally {
        activity.finished()
      }


    }
    else {
      initProjectsManager(false)
      projectsManager.addManagedFilesWithProfiles(files, MavenExplicitProfiles.NONE, null, null, true)
    }

    IndexingTestUtil.suspendUntilIndexesAreReady(project)
  }
```

When I reviewed the code for `importProjectsAsync()`, I noticed some changes in the method. I am attaching the code differences between versions `2024.1.7` and `2024.2`

<img width="1651" alt="Screenshot 2025-01-13 at 10 47 14 AM" src="https://github.com/user-attachments/assets/85a56478-4cc3-4714-af05-87cfc775b08a" />

There are two additional parameters present in the code built using version `2024.2`

**Comment by vaisakhkannan** [comment]
The change from the method above is due to the update in the IntelliJ version, and the Intellij team will not revert the code back to the previous version. Therefore, we need to find a way to resolve this unit test failure on our end.

**Comment by dessina-devasia** [comment]
After IntelliJ's modification to `importProjectsAsync`, which now includes `SimpleStructureProjectVisitor()`, we are encountering a **deadlock** when invoking this function from Java using BuildersKt.runBlocking.

```
BuildersKt.runBlocking(
              EmptyCoroutineContext.INSTANCE,
              (scope, continuation) -> importProjectsAsync(pomFiles.toArray(VirtualFile[]::new), continuation)
      );
```

The issue appears to have originated from IntelliJ’s side, likely due to changes in indexing behavior after the modification that introduced `SimpleStructureProjectVisitor()`. The deadlock occurs when calling `importProjectsAsync` from Java using `BuildersKt.runBlocking`, as the function waits on `IndexingTestUtil.suspendUntilIndexesAreReady`, which in turn requires `coroutine` execution but gets blocked. This was not an issue before the IntelliJ modification, suggesting that recent indexing changes may have affected coroutine execution.

**Observed Behavior**

- The function **never completes**, causing the thread to hang indefinitely.
- **Thread dump** indicates that the coroutine is waiting on `IndexingTestUtil.suspendUntilIndexesAreReady`, but due to `runBlocking`, the required coroutine execution is blocked.
- This issue was **not present before the IntelliJ modification**.

**Possible Cause**

- `runBlocking` is a synchronous call, meaning it holds onto a thread until the coroutine finishes.
- The inclusion of `SimpleStructureProjectVisitor()` appears to have altered internal execution flow, causing a circular wait scenario that leads to deadlock.

**Action Plan**

- Highlight the issue in the JetBrains channel, as their Slack channel has moved to another platform, and raise the issue there as well.
- Park this issue while we await a fix from the IntelliJ team.
- Prioritize issue #1226 as high priority since this issue is currently on hold.

**Comment by vaisakhkannan** [comment]
I have started a conversation on the new JetBrains forum since slack is no longer available to ask questions. Here is the link.

https://platform.jetbrains.com/t/unit-test-failure-for-2024-2-intellij-version/368

**Comment by vaisakhkannan** [comment]
I have gone through this JetBrains [documentation](https://plugins.jetbrains.com/docs/intellij/using-kotlin.html#adding-kotlin-support) and tried `kotlin.stdlib.default.dependency = false`, but that didn't work. I also checked different versions of 'org.jetbrains.kotlin.jvm', but that didn't work either.

**Comment by vaisakhkannan** [comment]
Tried below codes

```
try {
    BuildersKt.runBlocking(
        EmptyCoroutineContext.INSTANCE,
        (scope, continuation) -> importProjectsAsync(pomFiles.toArray(VirtualFile[]::new), continuation)
    );
    System.out.println("importProjectsAsync completed.");
} catch (Exception e) {
    e.printStackTrace(); // Print any exception that occurs
}

```

Failure : ERROR: File accessed outside allowed roots: file:///private/var/folders/rj/c3d948vn3t974m16bnjg_sg80000gn/T/unitTest_generatedAnnotation_2syr0ou6D5u9uhUhhtv5n3h8r8c/GeneratedAnnotation_2syr0ww3OImd64M7TbfmB09w2sW/jakarta-sample0/pom.xml;

**Comment by vaisakhkannan** [comment]
Instead of the below code, I tried `importProjectsWithErrors(pomFiles.toArray(VirtualFile[]::new));`

```
BuildersKt.runBlocking(
                EmptyCoroutineContext.INSTANCE,
                (scope, continuation) -> importProjectsAsync(pomFiles.toArray(VirtualFile[]::new), continuation)
        );
```

The tests passed, and I manually tried to reproduce the test failure. Those are working.

**Comment by vaisakhkannan** [comment]
@TrevCraw,  I tested this fix, `importProjectsWithErrors(pomFiles.toArray(VirtualFile[]::new));`, in IntelliJ version 2024.1 with the built code, and in that case, all tests passed.

**Comment by vaisakhkannan** [comment]
As part of the investigation into the failed tests, let's take the example `io.openliberty.tools.intellij.lsp4jakarta.it.beanvalidation.BeanValidationTest#validFieldConstraints.`

I ran the test in IntelliJ version 2024.2, and the test failed due to an assertion failure on the diagnostic. Below is a screenshot showing the difference in the diagnostic.

<img width="1637" alt="Image" src="https://github.com/user-attachments/assets/116a13b7-ee68-4f2f-bd5e-d06c2a6e5c3a" />

Based on the [code](https://github.com/OpenLiberty/liberty-tools-intellij/blob/609f5b6e41b2a4ddf2e6d2f5356bf375771e68ce/src/test/java/io/openliberty/tools/intellij/lsp4jakarta/it/beanvalidation/BeanValidationTest.java#L54), we can see that there is no diagnostic for that case, which is expected, but in actual value, we are getting a diagnostic. I checked in both VSCode and Eclipse, and in both cases, the diagnostic is not present.

<img width="1165" alt="Image" src="https://github.com/user-attachments/assets/65b37dc7-977c-42fe-bcf8-ee022638d419" />

<img width="942" alt="Image" src="https://github.com/user-attachments/assets/d5158240-703a-4988-b113-1c25cbebade1" />

But one important thing I found is that I ran the extension in IntelliJ, opened the `Jakarta sample`, and the `ValidConstraints` class. However, there are no diagnostics, yet unfortunately, our test is failing.

<img width="1387" alt="Image" src="https://github.com/user-attachments/assets/a1d579f9-735b-4121-a381-d3e3baa62ddc" />

 The same fix mentioned above works fine in `2024.1`(All unit tests are passing). I just assume that something has changed in 2024.2, causing the failure of 5 tests or the fix we have,` importProjectsWithErrors(pomFiles.toArray(VirtualFile[]::new));`, is not adequate.

What do you think @TrevCraw, @mrglavas ?

**Comment by mrglavas** [comment]
> The same fix mentioned above works fine in `2024.1`(All unit tests are passing). I just assume that something has changed in 2024.2, causing the failure of 5 tests or the fix we have,` importProjectsWithErrors(pomFiles.toArray(VirtualFile[]::new));`, is not adequate.
> 
> What do you think [@TrevCraw](https://github.com/TrevCraw), [@mrglavas](https://github.com/mrglavas) ?

@vaisakhkannan Have you stepped through one of these test cases in a debugger to see where the differences between 2024.1 and 2024.2 are? Presumably you should be able to observe the point where the program state / events start diverging and that might provide some clues as to the root cause.

**Comment by vaisakhkannan** [comment]
@mrglavas 
I found the problem in the test method `io.openliberty.tools.intellij.lsp4jakarta.it.beanvalidation.BeanValidationTest#validFieldConstraints` 

The value from `type.getCanonicalText()` is `java.math.BigDecimal` in IntelliJ 2024.1, causing the condition to fail, which leads to not adding the diagnostic.

<img width="1645" alt="Image" src="https://github.com/user-attachments/assets/433a1c3d-5575-48a9-a9d5-c9c0aa6947cf" />

The value from `type.getCanonicalText()` is `BigDecimal` in IntelliJ 2024.2, causing the condition to get success, which leads to  adding the diagnostic.

<img width="1654" alt="Image" src="https://github.com/user-attachments/assets/dca234e5-d93d-40f0-9e84-2207ace4a581" />

Our [code](https://github.com/OpenLiberty/liberty-tools-intellij/blob/86e9a9b7407a6df39c3368f144bf77af8f12812a/src/main/java/io/openliberty/tools/intellij/lsp4jakarta/lsp4ij/beanvalidation/BeanValidationConstants.java#L65) String should change to `BigDecimal` only when we update platform version as 2024.2.

Failure on test `io.openliberty.tools.intellij.lsp4jakarta.it.persistence.JakartaPersistenceTest#removeFinalModifiers` is also similar issue

<img width="1551" alt="Image" src="https://github.com/user-attachments/assets/06a8b1d4-5709-4895-ace1-e885e15c2dca" />

**Comment by vaisakhkannan** [comment]
Based on the latest JetBrains [reply](https://platform.jetbrains.com/t/difference-in-the-value-returned-by-the-getcanonicaltext-method-when-the-code-is-built-using-intellij-version-2024-2/544/14?u=vaisakh_t) in the forum, I tested the same scenarios by changing the `JDK` to versions `11` and `21` with IntelliJ version `2024.1` and `2024.2`. 

Following their latest suggestion with intellij version `2024.1`, I evaluated the value of `PsiUtil.resolveClassInClassTypeOnly(type)`. The value was `not null`, but the test failed due to the following error:

`Leaked SDKs: [java 11 Version: java 11 Path: (/Users/home/.m2/repository/org/jetbrains/mockjdk/mockjdk-base-java/11.0)].Please remove leaking SDKs by using ProjectJdkTable.getInstance().removeJdk() or by disposing of the ProjectJdkImpl.Registration trace for 'java 11' is shown as the cause.java.lang.AssertionError: Leaked SDKs: [java 11 Version: java 11 Path: (/Users/home/.m2/repository/org/jetbrains/mockjdk/mockjdk-base-java/11.0)]`.

With IntelliJ version `2024.2` and the same suggested approach shared by the jetbrain team, I evaluated `PsiUtil.resolveClassInClassTypeOnly(type)`, but this time the value was `null`, and the test still failed with the same error we are currently facing. It seems that nothing is working for IntelliJ version `2024.2` at the moment.

I also tested other scenarios with IntelliJ 2024.2 that is changing the `jdk` versions of `Jakarta-sample` project and language level that specified in the `pom.xml` file, and the value of the following expression is always null:

`PsiUtil.resolveClassInClassTypeOnly(type)`
Somehow, we need this value to be `not null` ( actual `Psitype` value)  for the tests to pass.
